src/vs/workbench/contrib/chat/browser/chatPasteProviders.ts

```diff
@@ -136,6 +163,119 @@ export function isImage(array: Uint8Array): boolean {
+		customDataTransfer.append('application/vnd.code.additional-editor-data', createStringDataTransferItem(rangesString));
+		return customDataTransfer;
+	}
+}
+
+export class PasteTextProvider implements DocumentPasteEditProvider {
+
+	public readonly kind = new HierarchicalKind('chat.attach.text');
+	public readonly pasteMimeTypes = ['application/vnd.code.additional-editor-data'];
+
+	constructor(
+		private readonly chatWidgetService: IChatWidgetService
+	) { }
+
+	async provideDocumentPasteEdits(model: ITextModel, ranges: readonly IRange[], dataTransfer: IReadonlyVSDataTransfer, context: DocumentPasteContext, token: CancellationToken): Promise<DocumentPasteEditsSession | undefined> {
+		if (model.uri.scheme !== ChatInputPart.INPUT_SCHEME) {
+			return;
+		}
+		const text = dataTransfer.get(Mimes.text);
+		const editorData = dataTransfer.get('vscode-editor-data');
+		const additionalEditorData = dataTransfer.get('application/vnd.code.additional-editor-data');
+
+		if (!editorData || !text || !additionalEditorData) {
+			return;
+		}
+
+		const textdata = await text.asString();
+		const metadata = JSON.parse(await editorData.asString());
+		const additionalData = JSON.parse(await additionalEditorData.asString());
+
+		let widget = this.chatWidgetService.getWidgetByInputUri(model.uri);
+		if (!widget) {
+			return;
+		}
+
+		const copiedContext = await getCopiedContext(textdata, additionalData.uri, metadata.mode, additionalData.ranges);
+
+		if (token.isCancellationRequested || !copiedContext) {
+			return;
+		}
+
+		const currentContextIds = widget.attachmentModel.getAttachmentIDs();
+		if (currentContextIds.has(copiedContext.id)) {
+			return;
+		}
+
+		const customEdit = {
+			resource: model.uri,
+			variable: copiedContext,
+			undo: () => {
+				widget = this.chatWidgetService.getWidgetByInputUri(model.uri);
+				if (!widget) {
+					throw new Error('No widget found for undo');
+				}
+				widget.attachmentModel.delete(copiedContext.id);
+			},
+			redo: () => {
+				widget = this.chatWidgetService.getWidgetByInputUri(model.uri);
+				if (!widget) {
+					throw new Error('No widget found for redo');
+				}
+				widget.attachmentModel.addContext(copiedContext);
+			},
+			metadata: { needsConfirmation: false, label: copiedContext.name }
+		};
+
+		return {
+			edits: [{
+				insertText: '', title: 'Empty Edit', kind: new HierarchicalKind(Mimes.text), handledMimeType: Mimes.text,
+				additionalEdit: {
+					edits: [customEdit],
+				}
+			}],
+			dispose() { },
+		};
+	}
+}
+
+async function getCopiedContext(code: string, file: string, language: string, ranges: IRange): Promise<IChatRequestVariableEntry | undefined> {
+	const fileName = file.split('/').pop() || 'unknown file';
+	const start = ranges.startLineNumber;
+	const end = ranges.endLineNumber;
+	const resultText = `Copied Selection of Code: \n\n\n From the file: ${fileName} From lines ${start} to ${end} \n \`\`\`${code}\`\`\``;
```

Comment:
Does this need to be valid markdown? If it does, you can use the `createFencedCodeBlock` helper with have for copilot as this will handle cases where the code contains triple backticks